File: backend/app/schemas/aid_request.py
"""AidRequest schema module"""

import datetime
from typing import Optional
from app.utils import AidRequestStatus
from pydantic import BaseModel


class AidRequestSchema(BaseModel):
    """
    The schema used to validate and structure data
    related to aid requests
    """
    id: int
    name: str
    description: str
    image: str
    location: str
    status: str
    deadline: datetime.datetime

    soldier_id: int
    volunteer_id: int | None
    category_id: int | None

    class Config:
        orm_mode = True


class AidRequestSchemaIn(BaseModel):
    """
    The schema used to validate and structure data
    related to aid requests
    """

    name: str
    description: str
    image: str
    location: str
    deadline: datetime.datetime

    volunteer_id: int | None
    category_id: int | None

class AidRequestSchemaInWithoutVolId(BaseModel):
    name: str
    description: str
    image: str
    location: str
    deadline: datetime.datetime
    category_id: int

    class Config:
        orm_mode = True

class AidRequestCreateMultipart(BaseModel):
    name: str
    description: str
    location: str
    deadline: datetime.datetime
    category_id: int

class AidRequestSchemaUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    image: Optional[str] = None
    status: Optional[str] = None
    location: Optional[str] = None
    deadline: Optional[datetime.datetime] = None
    volunteer_id: Optional[int] = None
    category_id: Optional[int] = None

    class Config:
        arbitrary_types_allowed = True
        from_attributes = True

class AidRequestAssignStatus(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    image: Optional[str] = None
    location: Optional[str] = None
    deadline: Optional[datetime.datetime] = None
    volunteer_id: Optional[int] = None
    category_id: Optional[int] = None

    class Config:
        arbitrary_types_allowed = True
        from_attributes = True

File: frontend/frontend-fortis/src/app/schemas/aid-request.ts
export class AidRequest {
  id: number;
  name: string;
  description: string;
  image_path: string;
  deadline: Date;
  location: string;
  status: string;
  image!: string;
  soldier_id: number;
  volunteer_id: number;
  category_id: number;

  constructor(
    id: number,
    name: string,
    description: string,
    image_path: string,
    deadline: Date,
    location: string,
    status: string,
    image: string, 
    soldier_id: number,
    volunteer_id: number,
    category_id: number
  ) {
    this.id = id;
    this.name = name;
    this.description = description;
    this.image_path = image_path;
    this.deadline = deadline;
    this.location = location;
    this.status = status;
    this.image = image;
    this.soldier_id = soldier_id;
    this.volunteer_id = volunteer_id;
    this.category_id = category_id;
  }
}
